### PR TITLE
[dhctl] ModuleSource mirroring

### PR DESCRIPTION
**Pull request by mvasl**
## Description

Added dhctl subcommand that allows to pull `ModuleSource` modules to local filesystem and push them to third-party registry.

## Why do we need it, and what problem does it solve?

We already have Deckhouse mirroring, it's time for modules

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: dhctl
type: feature
summary: ModuleSource mirroring via `dhctl`.
impact_level: default
```

### CODE REVIEW

**Comment by mvasl** [comment]
Pull
Снимок экрана 2023-11-20 в 19 24 25

Push
Снимок экрана 2023-11-20 в 19 25 06

Pulled modules
Снимок экрана 2023-11-20 в 19 28 55

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/openstack'...

File: dhctl/pkg/app/mirror_modules.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+		if MirrorRegistry != "" {
+			_, err := url.Parse("docker://" + MirrorRegistry)
+			if err != nil {
+				return fmt.Errorf("Malformed registry URL: %w", err)
+			}
+		}
+
+		if MirrorModuleSourcePath == "" && MirrorRegistry == "" {
+			return errors.New("One of ----modules-source or --registry flags is required.")
```
Comment (name212) [inline dhctl/pkg/app/mirror_modules.go:72]:
```suggestion
			return errors.New("One of --modules-source or --registry flags is required.")
```

File: dhctl/pkg/operations/mirror/layouts.go [hunk 2]
```diff
@@ -64,32 +66,23 @@ func CreateOCIImageLayouts(
...
 	}
 	for layoutPtr, fsPath := range fsPaths {
-		if err := createEmptyImageLayoutAtPath(fsPath); err != nil {
-			return nil, fmt.Errorf("create OCI Image Layout at %s: %w", fsPath, err)
-		}
-		*layoutPtr, err = layout.FromPath(fsPath)
+		*layoutPtr, err = CreateEmptyImageLayoutAtPath(fsPath)
 		if err != nil {
-			return nil, fmt.Errorf("get OCI Image Layout from %s: %w", fsPath, err)
+			return nil, fmt.Errorf("create OCI Image Layout at %s: %w", fsPath, err)
```
Comment (name212) [inline dhctl/pkg/operations/mirror/layouts.go:71]:
```suggestion
			return nil, fmt.Errorf("Create OCI Image Layout at %s: %w", fsPath, err)
```

**Review by name212 — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 3]
```diff
@@ -445,6 +445,60 @@ Thus, Deckhouse images will be available at `https://your-harbor.com/d8s/deckhou
...
+   docker run -ti --pull=always -v $(HOME)/d8-modules:/tmp/d8-modules -v $(HOME)/module_source.yml:/tmp/module_source.yml registry.deckhouse.io/deckhouse/ce/install:v1.56.0 bash
+   ```
+
+   Please note the volumes mounted from the host file system into the installer container: a directory into which module images will be loaded, and the ModuleSource YAML manifest, which describes the source of third-party modules.
+
+1. Download module images from their source registry, described as a ModuleSource resource, into a dedicated directory using the command `dhctl mirror-modules`.
+
+   `dhctl mirror-modules` downloads only versions of modules available in the module release channels at the time of copying.
+
+   The following command will download module images from the source described in the ModuleSource resource located in the `$HOME/module_source.yml` file:
```
Comment (kirkonru) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:462]:
```suggestion
1. Pull module images from their source registry, defined as a ModuleSource resource, into a dedicated directory using the command `dhctl mirror-modules`.

   `dhctl mirror-modules` pulls only versions of modules available in the module release channels at the time of copying.

   The following command will pull module images from the source described in the ModuleSource resource located in the `$HOME/module_source.yml` file:
```

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 4]
```diff
@@ -445,6 +445,60 @@ Thus, Deckhouse images will be available at `https://your-harbor.com/d8s/deckhou
...
+
+   ```shell
+   dhctl mirror-modules -d /tmp/d8-modules -m /tmp/module_source.yml
+   ```
+
+1. Optional: Copy the `dhctl` binary from the container to the directory where Deckhouse images were pulled.
```
Comment (kirkonru) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:468]:
```suggestion
1. Optional: Copy the `dhctl` binary from the container to the directory to which Deckhouse images were pulled.
```

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 5]
```diff
@@ -445,6 +445,60 @@ Thus, Deckhouse images will be available at `https://your-harbor.com/d8s/deckhou
...
+
+   ```shell
+   cp /usr/bin/dhctl /tmp/d8-images/dhctl
+   ```
+
+1. To continue with installation, use the `dhctl` binary you copied before OR repeat steps 1 and 2 for the Deckhouse installer on the host with access to the air-gapped registry. Make sure the directory with the pulled modules images is mounted into the container.
```
Comment (kirkonru) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:474]:
```suggestion
1. To continue with installation, use the `dhctl` binary you copied earlier OR repeat steps 1 and 2 for the Deckhouse installer on the host with access to the air-gapped registry. Make sure the directory with the pulled modules images is mounted into the container.
```

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 6]
```diff
@@ -445,6 +445,60 @@ Thus, Deckhouse images will be available at `https://your-harbor.com/d8s/deckhou
...
+   ```
+
+   > Please note that the images will be uploaded to the registry along the path specified in the `--registry` parameter (in the example above - /deckhouse-modules).
+   > Before running the command, make sure this path exists in your registry, and the account you are using has write permissions.
+
+   If your registry does not require authentication, omit both `--registry-login` and `--registry-password` flags.
+
+1. After uploading the images to the air-gapped registry, edit the ModuleSource YAML manifest:
+
+    * Change the `.spec.registry.repo` field to the address that you specified in the `--registry` parameter when uploading images;
```
Comment (kirkonru) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:491]:
```suggestion
    * Change the `.spec.registry.repo` field to the address that you specified in the `--registry` parameter when you uploaded the images;
```

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 7]
```diff
@@ -445,6 +445,60 @@ Thus, Deckhouse images will be available at `https://your-harbor.com/d8s/deckhou
...
+    * Change the `.spec.registry.dockerCfg` field to a base64 string with the authorization data for your registry in `dockercfg` format. Refer to your registry's documentation for information on how to obtain this token.
+
+1. Apply the ModuleSource manifest received in the previous step to the cluster.
```
Comment (kirkonru) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:494]:
```suggestion
1. Apply the ModuleSource manifest you got in the previous step to the cluster.
```

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 8]
```diff
@@ -445,6 +445,60 @@ Thus, Deckhouse images will be available at `https://your-harbor.com/d8s/deckhou
...
+
+   ```shell
+   kubectl apply -f $HOME/module_source.yml
+   ```
+
+   Once the manifest has been applied, the modules are ready to use, please refer to the module developer documentation for further instructions on how to configure and use them.
```
Comment (kirkonru) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:500]:
```suggestion
   Once the manifest has been applied, the modules are ready for use. For more detailed instructions on configuring and using modules, please refer to the module developer's documentation.
```

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 9]
```diff
@@ -445,6 +445,60 @@ Thus, Deckhouse images will be available at `https://your-harbor.com/d8s/deckhou
...
+
+   ```shell
+   cp /usr/bin/dhctl /tmp/d8-images/dhctl
+   ```
+
+1. To continue with installation, use the `dhctl` binary you copied before OR repeat steps 1 and 2 for the Deckhouse installer on the host with access to the air-gapped registry. Make sure the directory with the pulled modules images is mounted into the container.
+
+1. Upload module images to the isolated registry using the `dhctl mirror-modules` command.
+
+   Example command for downloading images from the `/tmp/d8-modules` directory:
```
Comment (kirkonru) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:478]:
```suggestion
   Below is an example of a command for pulling images from the `/tmp/d8-modules` directory:
```

**Review by kirkonru — APPROVED** [review]

**Review by z9r5 — APPROVED** [review]

**Pull request merged**